cmd/sample-service/main.go

```diff
@@ -0,0 +1,178 @@
+	dbStore *store.Store
+}
+
+func (h *handler) list(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	users, err := h.dbStore.ListUsers(r.Context())
+	if err != nil {
+		writeErr(w, err)
+		return
+	}
+
+	data, err := json.Marshal(users)
+	if err != nil {
+		writeErr(w, err)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	w.Write(data)
+}
+
+func (h *handler) create(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		writeErr(w, err)
+		return
+	}
+
+	var user store.User
+	err = json.Unmarshal(body, &user)
+	if err != nil {
+		writeErr(w, err)
+		return
+	}
+
+	err = h.dbStore.CreateUser(r.Context(), user)
+	if err != nil {
+		writeErr(w, err)
+		return
+	}
+
+	w.WriteHeader(http.StatusCreated)
+	w.Write([]byte(`{"message": "user created"}`))
+}
+
+func main() {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Wait for an os.Interrupt signal
+	go waitForCtrlC(cancel)
+
+	// Creates a new Workload API client, connecting to provided socket path
+	// Environment variable `SPIFFE_ENDPOINT_SOCKET` is used as default
+	client, err := workloadapi.New(ctx, workloadapi.WithAddr(socketPath))
+	if err != nil {
+		log.Fatalf("Unable to create workload API client: %v", err)
+	}
+	defer client.Close()
+
+	x509Context, err := client.FetchX509Context(ctx)
+	if err != nil {
+		log.Fatalf("Unable to fetch x509Context %v", err)
+	}
+
+	db, err := common.NewMySQLDBWithSPIRETLSConfig(x509Context, mysqlUser, mysalDBName, "")
+	if err != nil {
+		log.Fatalf("Failed to create MySQL Client: %v", err)
+	}
+
+	h := &handler{
+		db:      db,
+		dbStore: store.New(db),
+	}
+
+	// Start X.509 watcher
+	go startWatcher(ctx, client, h)
+
+	log.Printf("Starting API handlers")
+	// Add API handlers
+	r := mux.NewRouter()
+	r.HandleFunc(usersAPIPath, h.list).Methods(http.MethodGet)
+	r.HandleFunc(usersAPIPath, h.create).Methods(http.MethodPost)
+	log.Fatal(http.ListenAndServe(":8888", r))
+}
+
+func startWatcher(ctx context.Context, client *workloadapi.Client, h *handler) {
+	// Start a watcher for X.509 SVID updates
+	doneCh := make(chan struct{}, 1)
+	go func() {
+		err := client.WatchX509Context(ctx, &x509Watcher{
+			h: h,
+		})
+		if err != nil && status.Code(err) != codes.Canceled {
+			log.Fatalf("Error watching X.509 context: %v", err)
+		}
+		doneCh <- struct{}{}
+	}()
+	<-doneCh
+}
+
+type x509Watcher struct {
+	mu sync.Mutex
+	h  *handler
+}
+
+// OnX509ContextUpdate is run every time an SVID is updated
+func (w *x509Watcher) OnX509ContextUpdate(c *workloadapi.X509Context) {
+	// Create new DB instance with udpate TLS config
+	db, err := common.NewMySQLDBWithSPIRETLSConfig(c, mysqlUser, mysalDBName, "")
+	if err != nil {
+		log.Printf("Failed to create MySQL Client: %v", err)
+		return
+	}
+
+	// Update DB instance in handler
+	w.mu.Lock()
+	w.h.db = db
```

Comment:
Don't we need to update the `*sql.DB` reference in the `*store.Store` rather than the `*handler`?

Reply:
Since they share the same pointer reference, updating one updates the other - however I agree this is confusing and `setDB()` approach sounds cleaner.